### PR TITLE
[jit] Implements divmod function

### PR DESCRIPTION
**Pull request by pandeykartikey**
This PR refer to issue #18627

### CODE REVIEW

File: torch/csrc/jit/register_prim_ops.cpp
Comment (wanchaol) [inline torch/csrc/jit/register_prim_ops.cpp]: please put this into `aten` namespace, everything that has schema should go to `aten` instead of `prim`, prim is reserved for unschematized operator, other ops like `abs` did it wrong and we will fix it in the future.

**Review by wanchaol — DISMISSED** [review]
Also, please add overloads for floating point numbers, this is now only working for integers which is not complete, consider using the micros (like `DEFINE_BINARY_OP`, etc) instead of adding it explicitly.

**Comment by pandeykartikey** [comment]
Hi @wanchaol the implementation of `floating point` and `integer` variation of divmod are very different. So, I have added it explicitly. 
There was also no micro which handles two outputs. Shall I add one micro for that too?

File: torch/csrc/jit/register_prim_ops.cpp
Comment (wanchaol) [inline torch/csrc/jit/register_prim_ops.cpp]: what about mixed operand types? like (int, float) and (float, int)? Can you also add these two? we should apply binary arithmetic operators rules for those use cases.

**Review by wanchaol — COMMENTED** [review]
looks good on the implementation, you don't need to add the micro for it considering that it is a rare use case that returns two results, making them separately sounds fine. it would be good if you can add mixed operand type overloads as well :)

File: test/test_jit.py
Comment (driazati) [inline test/test_jit.py]: Can you also check with some negative inputs? Python and C++ [differ in their mod with negative numbers](https://stackoverflow.com/questions/1907565/c-and-python-different-behaviour-of-the-modulo-operation), we want to make sure it matches the Python behavior

Comment (eellison) [reply test/test_jit.py]: Also need to error check for 0 input

File: torch/csrc/jit/register_prim_ops.cpp
Comment (pandeykartikey) [reply torch/csrc/jit/register_prim_ops.cpp]: `Apply binary arithmetic operators rules ` would it mean calculating quotient first using `/` and then calculating mod using it?

**Comment by ailzhang** [comment]
@pandeykartikey  would you mind resolving the comments & conflicts?

**Comment by pandeykartikey** [comment]
@ailzhang I was waiting for a reply by @wanchaol on a doubt about binary arithmetic on mixed operands. All the others comments have been resolved in the previous commit.

File: torch/csrc/jit/register_prim_ops.cpp
Comment (ailzhang) [reply torch/csrc/jit/register_prim_ops.cpp]: @pandeykartikey  Yep that's correct. In generally results returned here should match python `divmod` results in all combinations of operand types. Given there will be similar combo of them, it's the best to be generated (similar to `aten::log`) instead of writing each `Operator` separately with similar implementation.

**Comment by ezyang** [comment]
@wanchaol could you please follow up on this PR?

**Comment by pandeykartikey** [comment]
@ailzhang @wanchaol @driazati This PR seems complete from my side would you mind taking a look.

File: test/test_jit.py
Comment (wanchaol) [inline test/test_jit.py]: can't you just use `self.checkScript(func_int, (1024, 0))` here and other exception cases?

File: torch/csrc/jit/register_prim_ops.cpp
Comment (wanchaol) [inline torch/csrc/jit/register_prim_ops.cpp]: `\` in this macro are not aligned, can you indent and align them?

**Review by wanchaol — APPROVED** [review]
Thanks for adding the mix operand type overloads! The PR looks good to me, but it needs to rebase with the master since there're some file conflicts recently. Also have some small nits.

**Comment by pandeykartikey** [comment]
@wanchaol I have rebased the branch on master and made the other fixes.

**Comment by wanchaol** [comment]
@pandeykartikey do you mind fix the clang-tidy linting issue?

```
Traceback (most recent call last):
  File "tools/clang_tidy.py", line 295, in <module>
    main()
  File "tools/clang_tidy.py", line 291, in main
    print(run_clang_tidy(options, line_filters, files))
  File "tools/clang_tidy.py", line 189, in run_clang_tidy
    raise RuntimeError(message.format(output))
RuntimeError: Found clang-diagnostic-errors in clang-tidy output: /home/travis/build/pytorch/pytorch/torch/csrc/jit/register_prim_ops.cpp:2407:9: error: expected ')' [clang-diagnostic-error]
        "aten::divmod(int x, int y) -> (int, int)",
        ^
/home/travis/build/pytorch/pytorch/torch/csrc/jit/register_prim_ops.cpp:2406:13: note: to match this '('
    Operator(
            ^
/home/travis/build/pytorch/pytorch/torch/csrc/jit/register_prim_ops.cpp:2407:9: error: expected unqualified-id [clang-diagnostic-error]
        "aten::divmod(int x, int y) -> (int, int)",
        ^
/home/travis/build/pytorch/pytorch/torch/csrc/jit/register_prim_ops.cpp:2472:13: error: use of undeclared identifier 'simpleClassTypeArg' [clang-diagnostic-error]
           !simpleClassTypeArg(schema_args[0], class_type) ||
            ^
/home/travis/build/pytorch/pytorch/torch/csrc/jit/register_prim_ops.cpp:2473:13: error: use of undeclared identifier 'simpleClassTypeArg' [clang-diagnostic-error]
           !simpleClassTypeArg(schema_args[1], class_type) ||
```

Maybe worth setting it up locally if you wish, see [contributing doc](https://github.com/pytorch/pytorch/blob/master/CONTRIBUTING.md#running-clang-tidy)

**Comment by pandeykartikey** [comment]
@wanchaol Fixed the linting issue.

**Review by zou3519 — DISMISSED** [review]
Some of the tests are failing (click one of the red 'X's). We should fix those before landing.
```
Jun 07 13:09:48 ======================================================================
Jun 07 13:09:48 ERROR: test_divmod (__main__.TestScript)
Jun 07 13:09:48 ----------------------------------------------------------------------
Jun 07 13:09:48 Traceback (most recent call last):
Jun 07 13:09:48   File "test_jit.py", line 5666, in test_divmod
Jun 07 13:09:48     self.checkScriptRaisesRegex(func_int, (1024, 0), RuntimeError, "division by 0")
Jun 07 13:09:48   File "test_jit.py", line 2898, in checkScriptRaisesRegex
Jun 07 13:09:48     script(*inputs)
Jun 07 13:09:48   File "test_jit.py", line 5638, in func_int
Jun 07 13:09:48     return divmod(a, b)
Jun 07 13:09:48 ZeroDivisionError: integer division or modulo by zero
Jun 07 13:09:48 
Jun 07 13:09:48 ----------------------------------------------------------------------
Jun 07 13:09:48 Ran 1934 tests in 585.133s
Jun 07 13:09:48 
Jun 07 13:09:48 FAILED (errors=1, skipped=45, expected failures=1)
Jun 07 13:09:49 Traceback (most recent call last):
Jun 07 13:09:49   File "test/run_test.py", line 449, in <module>
Jun 07 13:09:49     main()
Jun 07 13:09:49   File "test/run_test.py", line 441, in main
Jun 07 13:09:49     raise RuntimeError(message)
Jun 07 13:09:49 RuntimeError: test_jit failed!
Jun 07 13:09:49 + cleanup
Jun 07 13:09:49 + retcode=1
Jun 07 13:09:49 + set +x
```

**Comment by pandeykartikey** [comment]
Hi @wanchaol I switched back to my previous implementation of zero division error because the `checkScriptWithRegex` was not catching the error in the case of division by zero it was just throwing it and hence the previous tests didn't pass.

**Pull request closed without merging**
